### PR TITLE
Add support of storing the caBundle in a secret

### PR DESCRIPTION
**Pull request by farcaller**
- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This allows to e.g. pull it from vault via external-secrets (as ES cannot write to a ConfigMap).

```release-note
Allow to use a Secret for the caBundle
```

### CODE REVIEW

**Comment by sayboras** [comment]
:wave: friendly ping if you have a chance to perform the above. Also, would you mind to rebase with main branch for latest CI configuration.

**Comment by farcaller** [comment]
@sayboras thanks for a ping! I got a bit distracted by my cilium setup breaking down in new and exciting ways but I'm back to this and I have a working solution that I just need to rewrite this nix into helm yamls: 😄 

```nix
{
      patchCM = object:
        let
          volumes = object.spec.template.spec.volumes;
          projected = builtins.filter (v: builtins.hasAttr "projected" v) volumes;
          nonProjected = builtins.filter (v: builtins.hasAttr "projected" v == false) volumes;
          reprojected = builtins.map
            (p: lib.recursiveUpdate p (
              let
                caSources = builtins.filter (s: (s.configMap or { name = ""; }).name == "cilium-ca") p.projected.sources;
                nonCaSources = builtins.filter (s: (s.configMap or { name = ""; }).name != "cilium-ca") p.projected.sources;
                patchedSources = builtins.map
                  (s: {
                    secret = {
                      items = [{ key = "certificate"; path = (builtins.head s.configMap.items).path; }];
                      name = "cilium-ca";
                    };
                  })
                  caSources;
              in
              {
                projected.sources = nonCaSources ++ patchedSources;
              }
            ))
            projected;
        in
        (lib.recursiveUpdate object {
          spec.template.spec.volumes = nonProjected ++ reprojected;
        });

      patchOn = cond: object:
        if (cond object) then patchCM object
        else object;

      patchCiliumDS = patchOn (object: (object.kind == "DaemonSet" && object.metadata.name == "cilium"));
      patchApiServerDeployment = patchOn (object: (object.kind == "Deployment" && object.metadata.name == "clustermesh-apiserver"));
      patchHubbleRelayDeployment = patchOn (object: (object.kind == "Deployment" && object.metadata.name == "hubble-relay"));
      patchHubbleUIDeployment = patchOn (object: (object.kind == "Deployment" && object.metadata.name == "hubble-ui"));
}
```

**Comment by joamaki** [comment]
Heading on vacation so unassigned myself and picked @tommyp1ckles from sig-k8s by lottery (`sort -R sig-k8s | head -1`).

**Comment by joestringer** [comment]
cc @cilium/sig-hubble I think some of you folks have spent more time with Cilium's secret management previously and could potentially help advise on the solution here.

**Comment by gandro** [comment]
Also cc @giorio94 who added the CA bundle feature.

To me, the overall motivation "this allows to e.g. pull it from vault via external-secrets (as ES cannot write to a ConfigMap)." sounds very reasonable. Since we have not shipped the CA bundle feature in a release yet, I'd even say let's always make it a secret if that makes the CA bundle easier to manage (and I'm not a big fan of having a resource either be a configmap or a secret based on a Helm value).

It was also the case that before we added the CA bundle, the CA was already part of a secret, so that should also not be a problem. But I think it would be interesting to know from @giorio94 if there is a particular reason why the CA bundle is a ConfigMap, or if the intent just was that "it's a public key, so it doesn't need to be a secret".

I'm marking this as a release blocker for now, since I think this is something that we can quite easily resolve now, where it will become harder to resolve once we release the CA bundle feature.

**Comment by giorio94** [comment]
> It was also the case that before we added the CA bundle, the CA was already part of a secret, so that should also not be a problem. But I think it would be interesting to know from @giorio94 if there is a particular reason why the CA bundle is a ConfigMap, or if the intent just was that "it's a public key, so it doesn't need to be a secret".

No specific reason beside the fact that it is not a "secret" value as you mentioned, and to mimic the kubernetes behavior (specifically the `kube-root-ca.crt` ConfigMap, indeed also the default name was picket to look like that). But I don't personally have any objection to making that a secret if it simplifies the CABundle management process.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 1]
```diff
@@ -843,7 +843,7 @@ spec:
               - key: ca.crt
                 path: common-etcd-client-ca.crt
           {{- else }}
-          - configMap:
+          - {{ if .Values.tls.caBundle.useSecret }}secret{{ else }}configMap{{ end }}:
```
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: So there are multiple uses of `.Values.tls.caBundle.enabled` in our Helm charts. I think you will have to change all of them to use this pattern.

**Review by gandro — CHANGES_REQUESTED** [review]
Please run `make -C install/kubernetes`, it generates the `install/kubernetes/cilium/README.md` which needs to be updated if you add or change Helm values.

**Comment by gandro** [comment]
Thanks @giorio94 for the quick response! 

I did a quick look at trust-manager, and it seems to only support ConfigMaps unfortunately https://cert-manager.io/docs/projects/trust-manager/api-reference/#bundlespectarget 

So maybe we do need to use flag that switches between ConfigMap and Secrets :disappointed: 

In that case I'll remove the release-blocker label again, since the change with `useSecret` can be done in a backwards compatible manner and I think this PR still needs a bit of iteration.

**Comment by giorio94** [comment]
> I did a quick look at trust-manager, and it seems to only support ConfigMaps unfortunately https://cert-manager.io/docs/projects/trust-manager/api-reference/#bundlespectarget

Oh right, I actually forgot to double-check it. Thanks!

**Comment by farcaller** [comment]
(sorry for a sudden pull of all the people. I rebased from a shadow clone 😞 )

**Comment by gandro** [comment]
Another CI failure:

```
 Validating documentation (syntax, spelling)...

Please fix the following spelling mistakes:
* Documentation/helm-values.rst:2667: (useSecret)  tls.caBundle.useSecret
* Documentation/helm-values.rst:2667: (useSecret)  tls.caBundle.useSecret

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh useSecret 
```

**Review by chancez — COMMENTED** [review]
LGTM in terms of helm changes.

**Review by sayboras — APPROVED** [review]
The changes lgtm ✅

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 2]
```diff
@@ -845,7 +845,7 @@ spec:
...
```
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: nitpicking:
```suggestion
          - {{ .Values.tls.caBundle.useSecret | ternary "secret" "configMap" }}:
```

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 3]
```diff
@@ -2061,6 +2061,9 @@ tls:
     # -- Entry of the ConfigMap containing the CA trust bundle.
     key: ca.crt
 
+    # -- Use a Secret instead of a ConfigMap
```
Comment (kaworu) [inline install/kubernetes/cilium/values.yaml.tmpl]: nitpicking

```suggestion
    # -- Use a Secret instead of a ConfigMap.
```

**Review by kaworu — APPROVED** [review]
Thanks for the patch @farcaller!

Couple of cosmetic comments, but other than that LGTM.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 4]
```diff
@@ -845,7 +845,7 @@ spec:
               - key: ca.crt
                 path: common-etcd-client-ca.crt
           {{- else }}
-          - configMap:
+          - {{ .Values.tls.caBundle.useSecret | ternary "secret" "configMap" }}::
```
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: I think there is an extra column here
```suggestion
          - {{ .Values.tls.caBundle.useSecret | ternary "secret" "configMap" }}:
```

**Review by kaworu — CHANGES_REQUESTED** [review]
Please squash all the PR's commits into one on the way!

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml (hunk 4 above)
Comment (farcaller) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: Derp, thanks for catching it!

**Review by kaworu — APPROVED** [review]

**Comment by ti-mo** [comment]
One more: https://github.com/cilium/cilium/actions/runs/5354804806/jobs/9712313069?pr=25728
```
-     - Use a Secret instead of a ConfigMap
+     - Use a Secret instead of a ConfigMap.
```

**Comment by farcaller** [comment]
> One more

Done (thought that one was supposed to be re-generated from `make -C install/kubernetes`)

**Review by gandro — APPROVED** [review]

**Comment by ldelossa** [comment]
All required tests passing. Merging.

**Review by michi-covalent — APPROVED** [review]

**Review by qmonnet — APPROVED** [review]
“docs” change looks good. Thanks!

**Pull request merged**